File: src/container/EdgeCountOfBox.go
package container

import (
	"github.com/HuXin0817/dots-and-boxes/src/config"
	"github.com/HuXin0817/dots-and-boxes/src/model"
	"github.com/stretchr/testify/assert"
)

type EdgeCountOfBox [model.MaxBox]int

func (b *EdgeCountOfBox) Add(e model.Edge) (s int) {
	for _, box := range model.NearBoxes[e] {
		(*b)[box]++
		if config.DEBUG {
			assert.True(nil, (*b)[box] <= 4)
		}
		if (*b)[box] == 4 {
			s++
		}
	}
	return s
}

func (b *EdgeCountOfBox) MaxCount(e model.Edge) (c int) {
	for _, box := range model.NearBoxes[e] {
		c = max(c, b[box])
	}
	return c
}

File: src/model/EdgeCountOfBox.cpp
#include "EdgeCountOfBox.h"

int
EdgeCountOfBox::Add(Edge e) {
  int s = 0;
  for (auto box : EdgeBoxMapper::EdgeNearBoxes.At(e)) {
    At(box)++;
    assert(At(box) <= 4);
    if (At(box) == 4) {
      s++;
    }
  }
  return s;
}

int
EdgeCountOfBox::MaxCount(Edge e) const {
  int c = 0;
  for (auto box : EdgeBoxMapper::EdgeNearBoxes.At(e)) {
    c = std::max(c, At(box));
  }
  return c;
}
